vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,434 @@
+  self->m_connectPerformedPromise.set_value();
+  self->m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  auto self = shared_from_this();
+  try
+  {
+    co_await resume_background();
+
+    co_await resume_on_signal(self->m_connectPerformed.get());
+
+    if (self->m_readyState != ReadyState::Open)
+    {
+      self = nullptr;
+      co_return;
+    }
+
+    self->m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    self->m_writer.WriteBytes(arr);
+
+    co_await self->m_writer.StoreAsync();
+
+    if (self->m_pingHandler)
+    {
+      self->m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  auto self = shared_from_this();
+  if (self->m_writeQueue.empty())
+  {
+    self = nullptr;
+    co_return;
+  }
+
+  try
+  {
+    co_await resume_background();
+
+    co_await resume_on_signal(self->m_connectPerformed.get());  // Ensure connection attempt has finished
+
+    co_await resume_in_queue(self->m_dispatchQueue);            // Ensure writes happen sequentially
+
+    if (self->m_readyState != ReadyState::Open)
+    {
+      self = nullptr;
+      co_return;
+    }
+
+    size_t length;
+    std::pair<string, bool> front;
+    bool popped = self->m_writeQueue.try_pop(front);
+    if (!popped)
+    {
+      throw hresult_error(E_FAIL, L"Could not retrieve outgoing message.");
+    }
+
+    auto [message, isBinary] = std::move(front);
+
+    if (isBinary)
+    {
+      self->m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      self->m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      self->m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      //TODO: Use char_t instead of uint8_t?
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()),
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
+      self->m_writer.WriteBytes(arr);
+    }
+
+    co_await self->m_writer.StoreAsync();
+
+    if (self->m_writeHandler)
+    {
+      self->m_writeHandler(length);
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformClose()
+{
+  co_await resume_background();
+
+  co_await resume_on_signal(m_connectPerformed.get());
```

Comment:
It looks like I missed this on the other PR, but I think we need to keep a strong ptr here as well.

Reply:
Nevermind. I see that we wait on this to complete before destroying the object.